# CARGA DE NUESTROS DATOS

In [144]:
#Importamos Pandas

import pandas as pd

In [145]:
#Procedemos a leer nuestro dataframe y le asignamos el valor df

df = pd.read_csv('retail.csv')

df.head()

,Transaction ID,Date,Customer ID,Gender,Age,Product Category,Quantity,Price per Unit,Total Amount
0,1,2023-11-24,CUST001,Male,34,Beauty,3,50,150
1,2,2023-02-27,CUST002,Female,26,Clothing,2,500,1000
2,3,2023-01-13,CUST003,Male,50,Electronics,1,30,30
3,4,2023-05-21,CUST004,Male,37,Clothing,1,500,500
4,5,2023-05-06,CUST005,Male,30,Beauty,2,50,100


In [146]:
df.shape

(1000, 9)

In [147]:
#Procedemos a ver los tipos de datos de las columnas

print("--- TIPOS DE DATOS ---")
print(df.info())
print()

print("--- FILAS Y COLUMNAS ---")
print(df.shape)



--- TIPOS DE DATOS ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    1000 non-null   int64 
 1   Date              1000 non-null   object
 2   Customer ID       1000 non-null   object
 3   Gender            1000 non-null   object
 4   Age               1000 non-null   int64 
 5   Product Category  1000 non-null   object
 6   Quantity          1000 non-null   int64 
 7   Price per Unit    1000 non-null   int64 
 8   Total Amount      1000 non-null   int64 
dtypes: int64(5), object(4)
memory usage: 70.4+ KB
None

--- FILAS Y COLUMNAS ---
(1000, 9)


In [148]:
#Extraemos estadisticas descriptivas de nuestro Dataframe
df.describe().round(2)

,Transaction ID,Age,Quantity,Price per Unit,Total Amount
count,1000.00,1000.00,1000.00,1000.00,1000.0
mean,500.50,41.39,2.51,179.89,456.0
std,288.82,13.68,1.13,189.68,560.0
min,1.00,18.00,1.00,25.00,25.0
25%,250.75,29.00,1.00,30.00,60.0
50%,500.50,42.00,3.00,50.00,135.0
75%,750.25,53.00,4.00,300.00,900.0
max,1000.00,64.00,4.00,500.00,2000.0


In [149]:
#Buscamos filas duplicadas o datos vacios

print("--- DATOS VACÍOS ---")
print(df.isnull().sum())

print("\n--- FILAS DUPLICADAS ---")
print(df.duplicated().sum())

--- DATOS VACÍOS ---
Transaction ID      0
Date                0
Customer ID         0
Gender              0
Age                 0
Product Category    0
Quantity            0
Price per Unit      0
Total Amount        0
dtype: int64

--- FILAS DUPLICADAS ---
0


In [150]:
# BORRAMOS COLUMNAS INNECESARIAS

df.drop(columns=['Customer ID', 'Transaction ID'], inplace=True)

In [151]:
#Vemos nuestros datos nuevamente

df.head()

,Date,Gender,Age,Product Category,Quantity,Price per Unit,Total Amount
0,2023-11-24,Male,34,Beauty,3,50,150
1,2023-02-27,Female,26,Clothing,2,500,1000
2,2023-01-13,Male,50,Electronics,1,30,30
3,2023-05-21,Male,37,Clothing,1,500,500
4,2023-05-06,Male,30,Beauty,2,50,100


In [152]:
#Procedemos a crear el archivo de datos limpios

df.to_csv('Retail_Limpio.csv', index=False)

# FIJAMOS NUESTROS OBJETIVOS

## Impacto de las Caracteristicas del cliente en la Facturación

### Hipotesis

El perfil demográfico de los clientes influye significativamente en el volumen de facturación, existiendo segmentos específicos que generan un mayor ticket promedio que otros


# Objetivo 1: ¿Varía el ticket promedio significativamente entre hombres y mujeres?

In [153]:
#Arreglamos nuestros datos

Hombres = df[df['Gender'] == 'Male']

print('Total de hombres:', len(Hombres))
Hombres[['Date', 'Gender', 'Total Amount']].head(10)

Total de hombres: 490


,Date,Gender,Total Amount
0,2023-11-24,Male,150
2,2023-01-13,Male,30
3,2023-05-21,Male,500
4,2023-05-06,Male,100
6,2023-03-13,Male,50
7,2023-02-22,Male,100
8,2023-12-13,Male,600
10,2023-02-14,Male,100
11,2023-10-30,Male,75
12,2023-08-05,Male,1500


In [154]:
#Arreglamos nuestros datos

Mujeres = df[df['Gender'] == 'Female']

print('Total de Mujeres:', len(Mujeres))
print()
Mujeres[['Date', 'Gender', 'Total Amount']].head(10)



Total de Mujeres: 510



,Date,Gender,Total Amount
1,2023-02-27,Female,1000
5,2023-04-25,Female,30
9,2023-10-07,Female,200
14,2023-01-16,Female,2000
16,2023-04-22,Female,100
17,2023-04-30,Female,50
18,2023-09-16,Female,50
20,2023-01-14,Female,500
22,2023-04-12,Female,120
23,2023-11-29,Female,300


In [155]:
ticket_genero = df.groupby('Gender')['Total Amount'].mean().round(2)
print("--- Ticket Promedio por Género ---")
print(ticket_genero)

--- Ticket Promedio por Género ---
Gender
Female    456.55
Male      455.43
Name: Total Amount, dtype: float64


## CONCLUSION

### Tras analizar la base de datos transaccional, descubrimos que el ticket promedio no varía significativamente entre géneros (Mujeres: 456.55 vs Hombres: 455.43).

# Objetivo 2: ¿Cómo influye la edad de manera directa en la elección de la categoría del producto?


In [156]:

# Creamos una nueva columna
df['Grupo Etario'] = 'Mayor a 40'
df.loc[df['Age'] <= 40, 'Grupo Etario'] = '40 o Menor'

# Revisamos
print("Total de clientes por grupo:")
print(df['Grupo Etario'].value_counts())

Total de clientes por grupo:
Grupo Etario
Mayor a 40    534
40 o Menor    466
Name: count, dtype: int64


In [157]:
# Comparamos el Ticket Promedio
ticket_edades = df.groupby('Grupo Etario')['Total Amount'].mean().round(2)
print("--- Ticket Promedio por Edad ---")
print(ticket_edades)
print()

# Calculamos los porcentajes
pref_edades = df.groupby('Grupo Etario', as_index=False)['Product Category'].value_counts(normalize=True)
pref_edades['proportion'] = (pref_edades['proportion'] * 100).round(2)
print("--- Preferencias de Categoría ---")
print(pref_edades)

--- Ticket Promedio por Edad ---
Grupo Etario
40 o Menor    491.19
Mayor a 40    425.29
Name: Total Amount, dtype: float64

--- Preferencias de Categoría ---
  Grupo Etario Product Category  proportion
0   40 o Menor         Clothing       33.69
1   40 o Menor      Electronics       33.48
2   40 o Menor           Beauty       32.83
3   Mayor a 40         Clothing       36.33
4   Mayor a 40      Electronics       34.83
5   Mayor a 40           Beauty       28.84


# Objetivo 3 Mientras mayor edad tienen los clientes, hay tendencia a comprar artículos premium y a un mayor gasto total?

In [158]:
# CREAMOS RANGOS DE EDAD
df['Rango_Edad'] = 'Por asignar'

df.loc[df['Age'] <= 25, 'Rango_Edad'] = '18-25'
df.loc[(df['Age'] > 25) & (df['Age'] <= 35), 'Rango_Edad'] = '26-35'
df.loc[(df['Age'] > 35) & (df['Age'] <= 50), 'Rango_Edad'] = '36-50'
df.loc[df['Age'] > 50, 'Rango_Edad'] = '51+'

In [159]:
df['Tipo de Producto'] = 'Economico'
df.loc[df['Price per Unit'] > 300, 'Tipo de Producto'] = 'Premium'

df

,Date,Gender,Age,Product Category,Quantity,Price per Unit,Total Amount,Grupo Etario,Rango_Edad,Tipo de Producto
0,2023-11-24,Male,34,Beauty,3,50,150,40 o Menor,26-35,Economico
1,2023-02-27,Female,26,Clothing,2,500,1000,40 o Menor,26-35,Premium
2,2023-01-13,Male,50,Electronics,1,30,30,Mayor a 40,36-50,Economico
3,2023-05-21,Male,37,Clothing,1,500,500,40 o Menor,36-50,Premium
4,2023-05-06,Male,30,Beauty,2,50,100,40 o Menor,26-35,Economico
...,...,...,...,...,...,...,...,...,...,...
995,2023-05-16,Male,62,Clothing,1,50,50,Mayor a 40,51+,Economico
996,2023-11-17,Male,52,Beauty,3,30,90,Mayor a 40,51+,Economico
997,2023-10-29,Female,23,Beauty,4,25,100,40 o Menor,18-25,Economico
998,2023-12-05,Female,36,Electronics,3,50,150,40 o Menor,36-50,Economico


In [160]:
analisis = df.groupby(['Rango_Edad', 'Tipo de Producto']).agg({
    'Quantity': 'mean',
    'Total Amount': 'mean',
    'Price per Unit': 'mean'
})

print(analisis)

                             Quantity  Total Amount  Price per Unit
Rango_Edad Tipo de Producto                                        
18-25      Economico         2.393939    276.893939      105.568182
           Premium           2.594595   1297.297297      500.000000
26-35      Economico         2.664634    286.463415      103.750000
           Premium           2.512195   1256.097561      500.000000
36-50      Economico         2.488281    256.484375      101.933594
           Premium           2.596491   1298.245614      500.000000
51+        Economico         2.526104    238.192771       93.755020
           Premium           2.312500   1156.250000      500.000000


In [161]:
premium_por_edad = df[df['Price per Unit'] > 300].groupby('Rango_Edad').size()
economicos_por_edad = df[df['Price per Unit'] <= 300].groupby('Rango_Edad').size()


print('---CANTIDAD DE PRODUCTOS PREMIUM POR RANGO DE EDAD---')
print(premium_por_edad)
print()
print('---CANTIDAD DE PRODUCTOS ECONOMICOS POR RANGO DE EDAD---')
print(economicos_por_edad)

---CANTIDAD DE PRODUCTOS PREMIUM POR RANGO DE EDAD---
Rango_Edad
18-25    37
26-35    41
36-50    57
51+      64
dtype: int64

---CANTIDAD DE PRODUCTOS ECONOMICOS POR RANGO DE EDAD---
Rango_Edad
18-25    132
26-35    164
36-50    256
51+      249
dtype: int64


In [162]:
porcentaje_tipo = (df.groupby('Rango_Edad')['Tipo de Producto']
                     .value_counts(normalize=True)
                     .reset_index(name='Porcentaje'))

porcentaje_tipo['Porcentaje'] = (porcentaje_tipo['Porcentaje'] * 100).round(2)

porcentaje_tipo

,Rango_Edad,Tipo de Producto,Porcentaje
0,18-25,Economico,78.11
1,18-25,Premium,21.89
2,26-35,Economico,80.00
3,26-35,Premium,20.00
4,36-50,Economico,81.79
5,36-50,Premium,18.21
6,51+,Economico,79.55
7,51+,Premium,20.45


# Objetivo 4 ¿Qué rango de edad representa el mayor volumen de ingresos acumulados para el negocio?

In [163]:
analisis = df.groupby('Rango_Edad', as_index=False)['Total Amount'].sum()
analisis


,Rango_Edad,Total Amount
0,18-25,84550
1,26-35,98480
2,36-50,139660
3,51+,133310


# Objetivo 5 ¿El segmento que hace la mayor cantidad de transacciones es el mismo que deja la mayor cantidad de dinero?

In [164]:
transacciones = df.groupby(['Rango_Edad', 'Gender'], as_index=False).size()
transacciones

,Rango_Edad,Gender,size
0,18-25,Female,81
1,18-25,Male,88
2,26-35,Female,107
3,26-35,Male,98
4,36-50,Female,164
5,36-50,Male,149
6,51+,Female,158
7,51+,Male,155


In [165]:
ingresos = df.groupby(['Rango_Edad', 'Gender'], as_index=False)['Total Amount'].sum()
ingresos

,Rango_Edad,Gender,Total Amount
0,18-25,Female,39470
1,18-25,Male,45080
2,26-35,Female,55115
3,26-35,Male,43365
4,36-50,Female,70790
5,36-50,Male,68870
6,51+,Female,67465
7,51+,Male,65845


# PLOTLY

In [166]:
import plotly.express as px


# OBJETIVO 1

In [167]:
ticket_genero = df.groupby('Gender')['Total Amount'].mean().round(2).reset_index()

# Gráfico de barras
grafico1 = px.bar(ticket_genero, 
              x='Gender', 
              y='Total Amount', 
              color='Gender',           # ¡Esta es la palabra mágica para el color!
              text_auto='.2f',          # Ponemos el número para que se vea la diferencia
              title='Gasto Promedio por Género',
              color_discrete_map={'Female': 'lightpink', 'Male': 'darkblue'})
grafico1.show()

# OBJETIVO 2

In [168]:
df['Grupo de Edad'] =['Menor o igual a 40' 
                      if edad <= 40 
                      else 'Mayor a 40' 
                      for edad in df['Age']]

pref_edades = df.groupby('Grupo de Edad')['Product Category'].value_counts(normalize=True).reset_index(name='Porcentaje')

pref_edades['Porcentaje'] = (pref_edades['Porcentaje'] * 100).round(2)

pref_edades

,Grupo de Edad,Product Category,Porcentaje
0,Mayor a 40,Clothing,36.33
1,Mayor a 40,Electronics,34.83
2,Mayor a 40,Beauty,28.84
3,Menor o igual a 40,Clothing,33.69
4,Menor o igual a 40,Electronics,33.48
5,Menor o igual a 40,Beauty,32.83


In [169]:
df['Grupo Etario'] = 'Mayor a 40'

df.loc[df['Age'] <= 40, 'Grupo Etario'] = '40 o Menor'

pref_edades = df.groupby('Grupo Etario')['Product Category'].value_counts(normalize=True).reset_index()

pref_edades = pref_edades.rename(columns={'proportion': 'Porcentaje'})

pref_edades['Porcentaje'] = (pref_edades['Porcentaje'] * 100).round(2)

print(pref_edades)

  Grupo Etario Product Category  Porcentaje
0   40 o Menor         Clothing       33.69
1   40 o Menor      Electronics       33.48
2   40 o Menor           Beauty       32.83
3   Mayor a 40         Clothing       36.33
4   Mayor a 40      Electronics       34.83
5   Mayor a 40           Beauty       28.84


In [170]:
grafico2 = px.bar(
    pref_edades,
    x='Product Category',
    y='Porcentaje',
    color='Grupo Etario',
    barmode='group',
    text_auto='.2f',
    title='Preferencias por categoría según grupo de edad',
    labels={'Product Category': 'Categoría', 'proportion': 'Porcentaje (%)'}
)

grafico2.show()

# OBJETIVO 3

In [ ]:
grafico3 = px.bar(
    porcentaje_tipo,                   
    x='Rango_Edad',
    y='Porcentaje',
    color='Tipo de Producto',          
    text_auto='.2f',                   
    title='Proporción de compras: Económico vs Premium por Edad',
    color_discrete_map={'Economico': 'blue', 'Premium': 'red'},
    labels={'Rango_Edad': 'Rango de Edad', 'Tipo de Producto': 'Tipo de Producto'}
)

grafico3.show()

# OBJETIVO 4

In [ ]:
grafico4 = px.bar(
    analisis, 
    x='Rango_Edad', 
    y='Total Amount', 
    text_auto='.2s',            
    color='Rango_Edad',
    title='Ingresos Acumulados por Rango de Edad'
)

grafico4.show()

print("El rango de edad 36-50 representa el mayor volumen de ingresos acumulados.")


El rango de edad 36-50 representa el mayor volumen de ingresos acumulados.


# OBJ 5

In [173]:
grafico51 = px.bar(
    transacciones,
    x='Rango_Edad',
    y='size',
    color='Gender',
    barmode='group',
    title='Cantidad de transacciones por segmento',
    color_discrete_map={
        'Female': 'Lightpink',
        'Male': 'Blue'
    }
)

grafico51.show()

In [174]:
grafico52 = px.bar(
    ingresos,
    x='Rango_Edad',
    y='Total Amount',
    color='Gender',
    barmode='group',
    title='Ingresos totales por segmento',
    color_discrete_map={
        'Female': 'Pink',
        'Male': 'Blue'
    }
)

grafico52.show()